packages/colorpickers/src/elements/ColorSwatch/index.tsx

```diff
@@ -0,0 +1,109 @@
+import PropTypes from 'prop-types';
+import { ThemeContext } from 'styled-components';
+import { Tooltip } from '@zendeskgarden/react-tooltips';
+import { useGrid } from '@zendeskgarden/container-grid';
+import { useId } from '@zendeskgarden/container-utilities';
+import { StyledSwatchButton, StyledCheckIcon, StyledCell, StyledColorSwatch } from '../../styled';
+
+export interface ILabeledColor {
+  value: string;
+  label: string;
+}
+
+export interface IColorSwatchProps {
+  /** Enables wrapped keyboard navigation */
+  isWrapped?: boolean;
+  /** Sets the two-dimension array of labeled HEX and RGB/A string colors */
+  colors: ILabeledColor[][];
+  /** Sets the focused row index in a controlled color swatch */
+  rowIndex?: number;
+  /** Sets the focused column index in a controlled color swatch */
+  colIndex?: number;
+  /** Sets the selected row index in a controlled color swatch */
+  selectedRowIndex?: number;
+  /** Sets the selected column index in a controlled color swatch */
+  selectedColIndex?: number;
+  /** Sets the default focused row index in a uncontrolled color swatch */
+  defaultRowIndex?: number;
+  /** Sets the default focused column index in a uncontrolled color swatch */
+  defaultColIndex?: number;
+  /** Sets the default selected row index in a uncontrolled color swatch */
+  defaultSelectedRowIndex?: number;
+  /** Sets the default selected column index in a uncontrolled color swatch */
+  defaultSelectedColIndex?: number;
+  /** Handles color swatch changes */
+  onChange?: (rowIndex: number, colIndex: number) => void;
+  /** Handles color swatch select event */
+  onSelect?: (rowIndex: number, colIndex: number) => void;
+}
+
+/**
+ * @extends HTMLAttributes<HTMLTableElement>
+ */
+export const ColorSwatch = forwardRef<HTMLTableElement, IColorSwatchProps>(
+  ({ colors, isWrapped, ...props }, ref) => {
+    const { rtl } = useContext(ThemeContext);
+    const { getGridCellProps } = useGrid({
+      rtl,
+      matrix: colors,
+      selection: true,
+      wrap: isWrapped,
+      idPrefix: useId(),
+      ...props
+    });
+
+    return (
+      <StyledColorSwatch ref={ref}>
+        <tbody>
+          {colors.map((row: ILabeledColor[], rowIdx: number) => (
+            <tr key={row[0].value}>
+              {row.map((color: ILabeledColor, colIdx: number) => {
+                const { label, value } = color;
+                const gridCellProps = getGridCellProps({
+                  colIdx,
+                  rowIdx,
+                  type: 'button'
+                });
+
+                return (
+                  <StyledCell key={value}>
+                    <Tooltip content={label}>
+                      <StyledSwatchButton backgroundColor={value} {...gridCellProps}>
+                        {gridCellProps['aria-selected'] ? <StyledCheckIcon color={value} /> : null}
+                      </StyledSwatchButton>
+                    </Tooltip>
+                  </StyledCell>
```

Comment:
This all makes sense now 🤯 .

> Each cell is either a DOM descendant of or owned by a row element and has one of the following roles:
> * gridcell if the cell does not contain column or row header information.

The `getGridCellProps` may need to be modified since it's returning both trigger + cell props. May want to return two prop getter functions:

1) `getGridCellProps` for the props `aria-selected`, `role` (can be knocked out via `null`)
2) `getTriggerProps` for the props `id`, `tabIndex`, `onClick`, `onFocus`, `onKeydown`, `type`

Reply:
Here are the changes I'm looking to make to the DOM structure:

1) No roles will be added for each cell. The role will be implicitly defined via the top-level `role="grid"` as shown by the examples @jzempel shared
2) [`aria-selected`](https://www.w3.org/TR/wai-aria-1.1/#aria-selected) will be applied on the cell element (`td`) which implicitly has a `role="gridcell"`

Reply:
Hopefully you can get this corrected without needing to revisit `react-containers`. YMMV. But I agree that the underlying container will need to be updated as you are suggesting (as a minor or major breaking change) – and then subsequently applied back to this component.